src/openlcb/IfTcpImpl.hxx

```diff
@@ -0,0 +1,553 @@
+        NodeID gateway_node_id, long long sequence, string *tgt)
+    {
+        bool has_dst = Defs::get_mti_address(msg.mti);
+        HASSERT(tgt);
+        string &target = *tgt;
+        target.assign(HDR_LEN + msg.payload.size() +
+                (has_dst ? MSG_ADR_PAYLOAD_OFS : MSG_GLOBAL_PAYLOAD_OFS),
+            '\0');
+        uint16_t flags = FLAGS_OPENLCB_MSG;
+        error_to_data(flags, &target[HDR_FLAG_OFS]);
+        unsigned sz = target.size() - HDR_SIZE_END;
+        target[HDR_SIZE_OFS] = (sz >> 16) & 0xff;
+        target[HDR_SIZE_OFS + 1] = (sz >> 8) & 0xff;
+        target[HDR_SIZE_OFS + 2] = sz & 0xff;
+        node_id_to_data(gateway_node_id, &target[HDR_GATEWAY_OFS]);
+        node_id_to_data(sequence, &target[HDR_TIMESTAMP_OFS]);
+        error_to_data(msg.mti, &target[HDR_LEN + MSG_MTI_OFS]);
+        node_id_to_data(msg.src.id, &target[HDR_LEN + MSG_SRC_OFS]);
+        if (has_dst)
+        {
+            node_id_to_data(msg.dst.id, &target[HDR_LEN + MSG_DST_OFS]);
+            memcpy(&target[HDR_LEN + MSG_ADR_PAYLOAD_OFS], msg.payload.data(),
+                msg.payload.size());
+        }
+        else
+        {
+            memcpy(&target[HDR_LEN + MSG_GLOBAL_PAYLOAD_OFS],
+                msg.payload.data(), msg.payload.size());
+        }
+    }
+
+    /// Guesses the length of a tcp message from looking at the prefix of the
+    /// payload.
+    ///
+    /// @param data is the prefix of the incoming stream.
+    /// @param len is the number of available bytes in this prefix.
+    /// @return -1 if there is not enough bytes in the prefix yet to know the
+    /// length. Otherwise the total number of bytes that constitute this
+    /// packet.
+    static int get_tcp_message_len(const void *data, size_t len)
+    {
+        if (len < HDR_SIZE_END)
+        {
+            return -1;
+        }
+        const uint8_t *src = static_cast<const uint8_t *>(data);
+        uint32_t sz = src[HDR_SIZE_OFS];
+        sz <<= 8;
+        sz |= src[HDR_SIZE_OFS + 1];
+        sz <<= 8;
+        sz |= src[HDR_SIZE_OFS + 2];
+        return sz + HDR_SIZE_END;
+    }
+
+    /// Parses a TCP message format (from binary payload) into a general
+    /// OpenLCB message.
+    /// @param src the rendered TCP message.
+    /// @param tgt the output generic message
+    /// @return false if the message is not well formatted or
+    /// it is not an OpenLCB message.
+    static bool parse_tcp_message(const string &src, GenMessage *tgt)
+    {
+        int expected_size = get_tcp_message_len(src.data(), src.size());
+        if (expected_size > (int)src.size() || expected_size < MIN_MESSAGE_SIZE)
+        {
+            LOG(WARNING, "Incomplete or incorrectly formatted TCP message.");
+            return false;
+        }
+        uint16_t flags = data_to_error(&src[0]);
+        if ((flags & FLAGS_OPENLCB_MSG) == 0)
+        {
+            return false;
+        }
+        HASSERT((flags & FLAGS_CHAINING) == 0);
```

Comment:
Should we have an @todo comment here since we will be removing chaining from the standards?

Reply:
this code assumes chaining has been removed, so we don't need to have a todo here in the code. I filed https://github.com/openlcb/documents/issues/10